libsignal-service-dotnet/websocket/WebSocketWrapper.cs

```diff
@@ -29,36 +29,96 @@ public class WebSocketWrapper
 
         public WebSocketWrapper(string uri, CancellationToken token)
         {
-            _ws = new ClientWebSocket();
-            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+            CreateSocket();
             _uri = new Uri(uri);
             Token = token;
         }
+        private void CreateSocket()
+        {
+            _ws = new ClientWebSocket();
+            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+        }
 
         public void HandleOutgoingWS()
         {
             Debug.WriteLine(TAG + "HandleOutgoingWS started");
+            byte[] buf = null;
             while (!Token.IsCancellationRequested)
             {
+                
                 try
                 {
-                    var buf = OutgoingQueue.Take(Token);
+                    if (buf == null)
+                        buf = OutgoingQueue.Take(Token);
+                    ReconnectTask?.Wait();
                     _ws.SendAsync(new ArraySegment<byte>(buf, 0, buf.Length), WebSocketMessageType.Binary, true, Token).Wait();
+                    buf = null; //set to null so we do not retry the same block
                 }
                 catch (TaskCanceledException e)
                 {
+                    if (!Token.IsCancellationRequested)
+                        Reconnect();
                     Debug.WriteLine(TAG + "HandleOutgoingWS shutting down");
                 }
                 catch (Exception e)
                 {
                     Debug.WriteLine(TAG + "WS SendAsync failed: " + e.Message);
-                    //TODO reconnect
+                    Reconnect();
                 }
             }
             //TODO dispose
             Debug.WriteLine(TAG + "HandleOutgoingWS finished");
         }
+        private Task ReconnectTask;
+        public void Reconnect()
+        {
+            if (Token.IsCancellationRequested)
+                return;
+            if (ReconnectTask != null)
+            {
+                ReconnectTask?.Wait();//race condition handling
```

Comment:
I just tested it in a .net core project:
The second ConnectAsync call triggers an InvalidOperationException, so the second calling task will be stuck since after each delay the task will fail to connect, and wait again.

Shouldn't the start of the function and the start of the lock be the same? I thought we might:
- rename _ws to something nice and make it volatile
- guard everything in Reconnect with a lock (SemaphoreSlim?)
- guarantee that Reconnect only returns when we have seen a working connection

This way
- the out and in tasks won't get broken references to the websocket
- if something goes wrong in any of them, they will call reconnect, block if neccessary (with a token) and have a pretty good chance to use a working socket in their next send/recv call.

If the state is returning bogus information, we can just keep track of the last successful send/recv/connect in a timestamp and react accordingly.